crate/client/src/findex_rest_client.rs

```diff
@@ -0,0 +1,129 @@
+        let server_url = format!("{}{}", self.rest_client.http_client.server_url, endpoint);
+
+        trace!(
+            "Initiating batch_read of {} addresses for index {} at server_url: {}",
+            addresses.len(),
+            self.index_id,
+            server_url
+        );
+
+        let response = self
+            .rest_client
+            .http_client
+            .client
+            .post(&server_url)
+            .body(Addresses::new(addresses).serialize()?)
+            .send()
+            .await?;
+
+        if !response.status().is_success() {
+            warn!("batch_read failed on server url {:?}.", server_url);
+            let err = handle_error(&endpoint, response).await?;
+            return Err(ClientError::RequestFailed(err));
+        }
+
+        let words: OptionalWords<WORD_LENGTH> =
+            OptionalWords::deserialize(&response.bytes().await?)?;
+
+        trace!(
+            "batch_read successful on server url {}. result: {}",
+            &server_url,
+            words
+        );
+
+        Ok(words.into_inner())
+    }
+
+    async fn guarded_write(
+        &self,
+        guard: (Self::Address, Option<Self::Word>),
+        tasks: Vec<(Self::Address, Self::Word)>,
+    ) -> Result<Option<[u8; WORD_LENGTH]>, ClientError> {
+        let endpoint = format!("/indexes/{}/guarded_write", self.index_id);
+        let server_url = format!("{}{}", self.rest_client.http_client.server_url, &endpoint);
+
+        trace!(
+            "Initiating guarded_write of {} values for index {} at server_url: {}",
+            tasks.len(),
+            self.index_id,
+            &server_url
+        );
+
+        // BEGIN TODO: using `Serializable` avoids re-coding vector
+        // concatenation. Anyway, this should be abstracted away in a function.
+        let guard_bytes = Guard::new(guard.0, guard.1).serialize()?;
+        let task_bytes = Tasks::new(tasks).serialize()?;
+        let mut request_bytes = Vec::with_capacity(guard_bytes.len() + task_bytes.len());
+        request_bytes.extend_from_slice(&guard_bytes);
+        request_bytes.extend_from_slice(&task_bytes);
+        // END TODO
+
+        let response = self
+            .rest_client
+            .http_client
+            .client
+            .post(&server_url)
+            .body(request_bytes)
+            .send()
+            .await?;
+
+        if !response.status().is_success() {
+            warn!("guarded_write failed on server url {}.", server_url);
+            let err = handle_error(&endpoint, response).await?;
+            return Err(ClientError::RequestFailed(err));
+        }
+
+        let guard = {
+            let bytes = response.bytes().await?;
+            let words: Vec<_> = OptionalWords::deserialize(&bytes)?.into();
+            words.first().copied().ok_or_else(|| {
+                ClientError::RequestFailed(format!(
+                    "Unexpected response from server. Expected 1 word, got {}",
```

Comment:
```suggestion
                    "Unexpected response from server. Expected 1 word, none given",
```

Then you can remove the copy.

Reply:
using your suggestion: `words.into_iter().next().ok_or_else(|| ...)`